File: src/main/java/cn/bromine0x23/plantuml/controllers/UmlController.java
package cn.bromine0x23.plantuml.controllers;

import cn.bromine0x23.plantuml.services.TranscoderService;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.CookieValue;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;

import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;

/**
 * @author <a href="mailto:bromine0x23@163.com">Bromine0x23</a>
 */
@Controller
@RequestMapping("uml")
public class UmlController {

	private final TranscoderService transcoderService;

	public UmlController(TranscoderService transcoderService) {
		this.transcoderService = transcoderService;
	}

	@GetMapping("{encodedSource:.*}")
	public String show(
		@PathVariable("encodedSource") String encodedSource,
		@CookieValue(value = "layout", required = false) String layout,
		Model model
	) {
		var source = transcoderService.decode(encodedSource);
		model.addAttribute("encodedSource", encodedSource);
		model.addAttribute("source", source);
		if ("left-right".equals(layout)) {
			return "editor/left-right";
		} else {
			return "editor/fluid";
		}
	}

	@PostMapping
	public String create(@RequestParam("source") String source) {
		return "redirect:uml/" + transcoderService.encode(source);
	}

	@PostMapping("recover")
	public String recover(@RequestParam("url") String url) throws IOException {
		return "redirect:" + extractEncodedSource(url);
	}

	private String extractEncodedSource(String url) throws MalformedURLException {
		var path = new URL(url).getPath();
		var segments = path.split("/");
		if (segments.length > 2) {
			return segments[segments.length - 1];
		}
		return "SyfFKj2rKt3CoKnELR1Io4ZDoSa70000";
	}

}

File: src/main/kotlin/cn/bromine0x23/plantuml/controllers/UmlController.kt
package cn.bromine0x23.plantuml.controllers

import cn.bromine0x23.plantuml.services.UmlService
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import java.net.URL

@Controller
@RequestMapping("uml")
class UmlController(
	private val umlService: UmlService
) : ApplicationController() {

	@GetMapping("{encodedSource:.*}")
	fun show(
		@PathVariable("encodedSource") encodedSource: String,
		model: Model
	): String {
		val source = umlService.decode(encodedSource)
		model.addAttribute("encodedSource", encodedSource)
		model.addAttribute("source", source)
		return "index"
	}

	@PostMapping
	fun create(
		@RequestParam("source") source: String
	): String {
		return "redirect:uml/" + umlService.encode(source)
	}

	@PostMapping("recover")
	fun recover(
		@RequestParam("url") url: String
	): String {
		return "redirect:${extractEncodedSource(url)}"
	}

	private fun extractEncodedSource(url: String): String {
		val path = URL(url).path
		val segments = path.split('/')
		if (segments.size > 2) {
			return segments.last()
		}
		return "SyfFKj2rKt3CoKnELR1Io4ZDoSa70000"
	}

}